### PR TITLE
Expose authentication methods on outbound federation

### LINKED ISSUE #1099: Expose authentication methods on outbound federation
**Is your feature request related to a problem? Please describe.**
In the scenario where PeeringDB is the [authorization server](https://openid.net/specs/openid-connect-basic-1_0.html#Terminology) used by third parties to signin PeeringDB users automatically, it is useful to know the successful authentication processes the user has met upon signin towards establishing a level of assurance of the identity.

The OIDC specification defines the `acr` and `amr` [claims](https://openid.net/specs/openid-connect-basic-1_0.html#IDToken) as two complementing elements to expose the identity assurance level and authentication methods that the user has met respectively upon signing in.

**Who is affected by the problem?**
Any third-party OIDC provider will benefit from this feature and will gain trust in the authentication process and subsequent operations allowed on their tooling.

**What is the impact?**
No operational changes to existing PeeringDB processes.

**Are there security concerns?**
No other security considerations other than those already in considered when implementing Oauth2/OIDC.

**Are there privacy concerns?**
The exposure of additional claims may reveal information about the user (ref [rfc8176#section-4](https://datatracker.ietf.org/doc/html/rfc8176#section-4)), however the changes proposed in this request are generic to not include personal identifiable information.

**Describe the solution you'd like**
There are standards (e.g. [PAPE](https://openid.net/specs/openid-provider-authentication-policy-extension-1_0.html), [RFC8176](https://datatracker.ietf.org/doc/html/rfc8176)) and proposals of standard ([EAP](https://openid.net/specs/openid-connect-eap-acr-values-1_0.html)) for each of them so this feature request is to propose to start with `amr` with values drawn from RFC8176 as a simpler mechanism to start with.

From that starting point, two use case scenarios could be delivered:
* a verified PeeringDB user has **no** FIDO/U2F second factor: the JWT id_token includes `amr: [pwd]`
* a verified PeeringDB user has FIDO/U2F enabled: the JWT id_token includes `amr: [pwd, hwk]`

**Do you think this feature will require a formal design?**
Possibly.

**Describe alternatives you've considered**
There is no known alternative available within the OIDC protocol to step up authentication once completed by PeeringDB.

**Could this feature request need support from the Admin Committee?**
Possibly.

**What is the proposed priority?**
Ideally to be completed together with #692 and/or #760

**Provide a rationale for any/all of the above**
PeeringDB is positioning itself not only as a supporting entity to coordinate peering practices but also as an element of trust within the participating networks. The more we can strengthen the security posture offered to those parties, the better trust and lower the barriers to run smooth operations for the benefit of our common customers.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by vegu**
will fix #1099, should be merged into support_202402 (#1564) (or into master after that's merged)

### CODE REVIEW

**Review by grizz — APPROVED** [review]

File: docs/dev/oauth.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+- `otp` - One-time password
+- `u2f` - Universal 2nd Factor (not in the RFC spec, but since we're not able to collect device attestation, we're using this as a proxy for security key usage - e.g., YubiKey, device pin, fingerprint scanner etc.)
+
+```python
+"amr": ["pwd", "mfa", "otp"] # password entered + OTP
+"amr": ["pwd", "mfa", "u2f"] # password entered + Security Key
+"amr": ["pwd"] # password entered
+"amr": ["mfa", "u2f", "otp"] # passwordless with security key + OTP
+"amr": ["mfa", "u2f", "u2f"] # passwordless with security key + plus 2fa with another security key
+"amr": ["u2f"] # password less without mfa
```
Comment (caguado) [inline docs/dev/oauth.md:59]: Passwordless is not `u2f` but technically is `webauthn` or `fido2` and is based on public keys.

File: peeringdb_server/views.py [hunk 2]
```diff
@@ -795,6 +803,18 @@ def view_profile_v1(request):
 
         data["networks"] = networks
 
+    # only add amr if amr scope is present
+    if scope_amr:
```
Comment (caguado) [inline peeringdb_server/views.py:807]: What's the reason to make the field opt-in via one additional scope?

This claim is already defined optional in OpenID Core spec for the JWT ID token so implementations will not be surprised by is addition.

File: docs/dev/oauth.md (hunk 1 above)
Comment (vegu) [reply docs/dev/oauth.md]: @caguado thanks for this - pushed update

File: mainsite/oauth2/validators.py [hunk 3]
```diff
@@ -30,3 +35,80 @@ def get_additional_claims(self):
...
+
+    def invalidate_authorization_code(self, client_id, code, request, *args, **kwargs):
+        print("invalidate_authorization_code")
+        grant = Grant.objects.get(code=code, application=request.client)
+        self.grant = grant
+        return super().invalidate_authorization_code(
+            client_id, code, request, *args, **kwargs
+        )
+
+    def _create_access_token(self, expires, request, token, source_refresh_token=None):
```
Comment (caguado) [inline mainsite/oauth2/validators.py:47]: Does this mean the `amr` won't be included in the JWT ID token?

The JWT ID token is sometimes used by the client (e.g. single-page application) to present personal information and when not reaching out to other OAuth2 resource servers but the client is a confidential one, it is all the info the application will consume so I would still add it to the JWT ID token too if it is not what this method does.

**Review by caguado — COMMENTED** [review]
Matt kindly asked to have a look so I hope my comments make sense. I have not deployed and not too familiar with the code anymore so I hope it makes sense.

File: docs/dev/oauth.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+- `mfa` - Multi-factor authentication
+- `otp` - One-time password
+- `u2f` - Universal 2nd Factor (not in the RFC spec, but since we're not able to collect device attestation, we're using this as a proxy for security key usage - e.g., YubiKey, device pin, fingerprint scanner etc.)
+
+```python
+"amr": ["pwd", "mfa", "otp"] # password entered + OTP
+"amr": ["pwd", "mfa", "u2f"] # password entered + Security Key
+"amr": ["pwd"] # password entered
+"amr": ["mfa", "u2f", "otp"] # passwordless with security key + OTP
+"amr": ["mfa", "u2f", "u2f"] # passwordless with security key + plus 2fa with another security key
```
[the remark this answers is not in the document]
Comment (vegu) [reply docs/dev/oauth.md]: I don't think we can currently differentiate, so putting `swk` seems reasonable to me.

File: peeringdb_server/views.py (hunk 2 above)
Comment (grizz) [reply peeringdb_server/views.py:807]: @caguado The idea was to let the user decide if they want to share that data with the party requesting it. Do you disagree?

File: docs/dev/oauth.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (vegu) [reply docs/dev/oauth.md]: i have pushed the change for this

File: mainsite/oauth2/validators.py (hunk 3 above)
Comment (vegu) [reply mainsite/oauth2/validators.py:39]: @caguado added to JWT ID token with https://github.com/peeringdb/peeringdb/pull/1565/commits/e27ab9dce69c8036262b27322a190a4f02889234

File: peeringdb_server/views.py [hunk 5]
```diff
@@ -3467,6 +3509,42 @@ def get_redirect_url(self):
...
+            #
+            # We'd want to differentiate fingerprint readers, iris scanners etc.
+            # but for that webauthn attestation is needed, which we currently
+            # do not collect.
+            #
+            # NOTE: by design if passwordless authentication was used
+            # it is required to be a different security key, so there may
+            # actually be cases of amr being "u2f u2f" which is accurate
+            # and RFC 8176 does not seem to disallow this (multiples of the same type).
+            amr.append("u2f")
```
[the remark this answers is not in the document]
Comment (vegu) [reply peeringdb_server/views.py]: @caguado i believe the change to `swk`  as discussed above, addresses this concern as well, so resolving this comment.

File: peeringdb_server/views.py (hunk 2 above)
Comment (caguado) [reply peeringdb_server/views.py:807]: I would disagree. From my interpretation of the [openid core spec](https://openid.net/specs/openid-connect-core-1_0.html#IDToken), this claim falls under the same core category as the subject, issuer, audience, issuance date, etc which are exposed with the `openid` scope and as they are part of the token attestation itself.

**Pull request merged**